### ISSUE TITLE
Securely exposing packages that leverage external connections

### ISSUE THREAD

**Issue by yakkomajuri** [PostHog/plugin-server#395]
Sparked by a discussion here: https://github.com/PostHog/plugin-repository/issues/11

We currently allow plugin users to leverage `fetch` and would also want to add support for libraries such as `pg`, which make connections via HTTP, TCP, etc.

However, as said by @mariusandra: "the connections will be made from our VPC, so they could theoretically find their way towards sensitive data".

While the issue came up when looking to add `pg`, @macobo pointed out that `fetch` is just as problematic: "...`clickhouse` exposes a HTTP api which can be used for evil things there." I guess the difference is there's no plugin that makes an HTTP request via `fetch` to an arbitrary host (not on Cloud at least), while plugins to send data to a Postgres instance for example would need to accept an arbitrary host.

## How to solve?

Well, this is the million dollar question. Some of this can be solved on an individual use-case basis, but it would be great to have a mechanism that applies to all plugins.

There's many possible levels to this:

### 1. One-off (individual plugin) basis

Ensuring all plugins that make it to Cloud have limitations on user inputs for connection parameters. This would be handled at the plugin level (e.g. doing a check on the host, port, etc.). This can be done easily now if we want to ship some plugins fast but is not a scalable approach.

### 2. Limitations at the library level

Another approach is to patch/fork the libraries we expose to plugin devs. This means they could safely take user input and make connections but the library would prevent anything that's dangerous. This is also not scalable if we want to do #286.

### 3. Server-level 

Ultimately (in an ideal world) we would want to get to a point where plugin devs can do as they wish, with all available packages, and we'd still be safe. This would require some configuration at the top level (preventing connections within the VPC). Quoting @macobo: "I guess next thing to dig into is how networking in node works in general and see whether there's a central hook to monkey-patch or figure out how to do some DNS trickery so plugins can't access instances."

## Discussion

So, what next? Do we want to start with 1 or 2 to ship certain plugins (e.g. Redshift)? Or do we want to take a step back and figure this all out at once?

**Comment by Twixes** [comment]
I'm pretty sure there's no way to do something bad network-side from a Node VM (aside from plain dumb denial of service). File system access, that's definitely a concern regarding libraries and something we must avoid. Arbitrary requests, should be safe.

**Comment by mariusandra** [comment]
Oh well, let's just add `node-postgres` then. 

In all likelihood, if someone gets our cloud `DATABASE_URL`, then trying to export events to it via a plugin is probably the least of our worries. They'll probably be in our systems and have much more than that.

If we control the plugins we have on cloud, we should be able to limit the damage from plugins. What I'm worried about is abuse of a postgres/redshift export plugin that also drop tables on setup if they don't match the schema or whatever. If we avoid that, we should be good.

Option 2 (patch these libraries to disallow our hosts) might be nice to add as well, though probably not [like this](https://github.com/PostHog/plugin-server/commit/b1ba65dbf6426fbe56861567662264ddb3eb1d49). That's too much maintenance overhead :D.

Option 3 won't work. We are talking about a scenario were we want to simultaneously allow and disallow TCP connections between one unix process and some servers. So the `pgPool` in `worker.ts` should connect to our hosts, but the `pgPool` in `vm.ts` (in the same nodejs process) should not. I don't think there's anything on the server level we can do that won't be a nightmare to support.

**Comment by yakkomajuri** [comment]
Yup, certainly think https://github.com/PostHog/plugin-server/pull/392 would be too much maintenance overhead. Goal there was to spark the discussion + move in an iterative manner and make sure we get Redshift out this sprint.

**Comment by macobo** [comment]
This is exactly the sort of thing that leads to larger vulnerabilities down the line.

:thinking: One idea that popped up was doing this via validating the address user passes as an argument to the plugin. 

Some things that need to be blocked to avoid risk of leakage:

- Disallow localhost
- Disallow raw IPs by default
- Disallow routes we have behind DNS (posthog.net)
- Disallow `amazonaws.com` urls (?)

Not sure on the last one - users could point their own DNS at things in amazonaws.

Also we could resolve DNS on the input and see whether it would fall within our VPC or not.

**Comment by yakkomajuri** [comment]
@macobo yup, that's what 1 (and to some extent 2) suggest. 

1 is a bit less scalable as it relies on us to edit community plugins that come in to check user inputs. 2 would do that at the lib level (or a wrapper around it), in a better way than https://github.com/PostHog/plugin-server/pull/392/commits/b1ba65dbf6426fbe56861567662264ddb3eb1d49

Can't disallow all `amazonaws.com` though as we have plugins for S3 and soon Redshift

**Comment by macobo** [comment]
I think you misunderstood.

My proposal is this. Say you have a plugin which accepts a URL as an argument when you install. For us to accept it into posthog cloud, you would need something like the following in `plugin.json`:

```json
{
  "name": "Sample plugin",
  "url": "https://github.com/PostHog/helloworldplugin",
  "description": "Greet the World and Foo a Bar, JS edition!",
  "main": "index.js",
  "config": [
    {
      "key": "url-with-port",
      "name": "Where to sync data",
      "type": "URLWithPort",
      "default": "your.domain:5432",
      "required": true
    }
  ]
}
```

`URLWithPort` is a custom type we provide which does the validations listed above on cloud. In code it would still appear as a string.

> Can't disallow all amazonaws.com though as we have plugins for S3 and soon Redshift

Just listing what needs to be blocked for this to be safe - AWS internal services (e.g. elasticache) are a target. For redshift or s3 we could whitelist them specifically, but `amazonaws.com` in general is suspect. :)

**Comment by yakkomajuri** [comment]
I see, but that doesn't solve the problem of it ultimately falling on us to verify plugin authors follow best practices (i.e. what ensures they'll use that instead of string?).

What about instead of modifying the actual lib like I did 😅 we change stuff like this:

```js
import * as nodeFetch from 'node-fetch'

export const imports = {
    // ...
    'node-fetch': fetch
    // ...
}
```

To:

```js
import nodeFetch from 'node-fetch'

const fetch = async (url: string, ...args: any) => {
     validateUrl(url)
     return (await nodeFetch(url, ...args))
}

export const imports = {
    // ...
    'node-fetch': fetch
    // ...
}
```

Would require some work for every package we introduce that makes network connections I guess which might make PostHog/plugin-server#286 hard, but will prevent us from having to mess with plugin devs' code.

**Comment by mariusandra** [comment]
I merged the updated node-postgres PR before seeing the full discussion here, apologies if that might seem rushed.

@macobo - I'm not sure what exactly "this" was here:
> This is exactly the sort of thing that leads to larger vulnerabilities down the line. 

The RedShiftPool or exposing the `pg` package to plugins?

Regarding config fields, some issues: we'd have to support a few different types to satisfy all needs, and it also pushes complexity from our implementation/plate onto the plugin authors. Perhaps we can avoid that.

Regarding patching all packages, we could do that, but it's a forever game of whack-a-mole. I don't mind some wrappers around more popular tools like `fetch` and `pg`, but what about `aws-sdk`? That's a behemoth. Version upgrades to any package will slow down if we go down this route as well.

We could expose a few "checkUrlForSuspiciousStuff" functions via plugin-contrib (this runs in the main thread scope, not in the vm), and make sure the postgres plugin calls it before passing on params. 

Then we add this as a requirement for a plugin to be allowed into the repository.

In any case, I think we're probably inflating the risk. As long as we limit and audit all plugin code that's installed on cloud, and don't allow plugins that can be used to automatically `drop tables` or `fetch post 100mb lol.png` anyone, we're good.

**Comment by yakkomajuri** [comment]
Speaking of wrappers for `fetch` and `pg`... https://github.com/PostHog/plugin-server/pull/397

I was writing this up and PostHog/plugin-server#392 got merged. **Definitely not scalable** to do this to all libs, but hey, let's us go back and forth on this thread here for a little longer :D

**Comment by macobo** [comment]
> @macobo - I'm not sure what exactly "this" was here:

Sorry - meant this issue as a whole. Most hacks I've read about start with I compromised this internal service to hop onto X and Y. The easiest way to "secure" these internal services is to keep them inaccessible from the outside world (e.g. in a VPC).

It's worth establishing a goal to evaluate a solution against. @yakkomajuri's PR had some pushback which might not be valid depending on the goal to optimize.

Proposal:
1. It should not be possible to access posthog-internal services _on cloud_ from within plugins
2. It should not be possible to use posthog _on cloud_ to DDOS another service

**Comment by Twixes** [comment]
I agree @macobo. I'd just like to avoid blanket bans, especially for trivially bypassable measures which would only frustrate users, so for 1. I think we should just make sure it's illegal to make requests to destinations resolving to internal IP addresses/blocks used by PostHog Cloud + loopback addresses.
For 2., we don't control network usage of plugin VMs, as they are built into the rest of the plugin server, but we could maybe make sure that at least `fetch()` has some rate limits (material for a separate PR).

**Comment by macobo** [comment]
> but we could maybe make sure that at least fetch() has some rate limits (material for a separate PR).

I think this is also covered by the vetting process.

> internal IP addresses/blocks used by PostHog Cloud + loopback addresses.

There is one thing this does not cover which I think might be targets - aws-internal services that are only accessible from the VPC. So things like elasticcache, etc which you'd access from `someservice.amazonaws.com`

**Comment by mariusandra** [comment]
I think it's decision time.

Adding some protection inside `fetch` makes sense, as it's a big wildcard that covers a lot of cases. However our best bet for cloud is to validate the URLs that come in to the plugins, inside all the plugins, and vet that every new/upgraded plugin we allow does this. This should probably be done with a function imported from `@posthog/plugin-contrib`.

If we try to come up with the perfect solution that patches all libraries we include, we'll be **forever** playing a game of whack-a-mole. For example, even with fetch it's easy to overlook HTTP redirects. What unknown unknowns will lurk for us inside the depths of the `aws-sdk` nodejs package? We will never know, and we don't have the resources to audit every nodejs package that we include because we will miss things; it happens to the best of us. Not to mention the maintenance cost when we need to upgrade to the next AWS SDK patch version.

Even if we do that, we won't catch everything, for example the aws-internal `someservice.amazonaws.com` services @macobo mentioned. We can check all plugin server `env`s for our own services (e.g. the graphile queue) and exclude those IPs, but where do you draw the line? Maybe not yet here, I don't know.

When we enable "write your own plugin with the cloud editor", we will need a radically different VM isolation approach with a very different networking layer anyway. I propose to push the search for the real solution for then.

Thus I propose that we just:
1) add host validity checks inside the `index.ts` of the pg/redshift/s3/snowflake/... plugins
2) get the `fetch` PR live, and add manual redirect handling
3) make sure all plugins we allow on cloud do this validity check

// Also, too bad we're using NodeJS 14, not 15, as otherwise we could use this to verify ip ranges: https://nodejs.org/api/net.html#net_blocklist_check_address_type

**Comment by yakkomajuri** [comment]
> get the fetch PR live, and add manual redirect handling

doing this now
